tests/test_dynamic_builtin_tools.py

```diff
@@ -0,0 +1,127 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+from pydantic_ai import Agent, RunContext
+from pydantic_ai.builtin_tools import AbstractBuiltinTool, WebSearchTool, WebSearchUserLocation
+from pydantic_ai.messages import ModelMessage, ModelResponse, TextPart
+from pydantic_ai.models import Model, ModelRequestParameters
+from pydantic_ai.settings import ModelSettings
+
+
+@dataclass
+class UserContext:
+    location: str | None
+
+
+async def prepared_web_search(ctx: RunContext[UserContext]) -> WebSearchTool | None:
+    if not ctx.deps.location:
+        return None
+
+    return WebSearchTool(
+        search_context_size='medium',
+        user_location=WebSearchUserLocation(city=ctx.deps.location),
+    )
+
+
+class InspectToolsModel(Model):
+    def __init__(self):
+        self.captured_tools: list[AbstractBuiltinTool] = []
+
+    @property
+    def model_name(self) -> str:
+        return 'inspect-tools-model'
+
+    @property
+    def system(self) -> str:
+        return 'test'
+
+    async def request(
+        self,
+        messages: list[ModelMessage],
+        model_settings: ModelSettings | None,
+        model_request_parameters: ModelRequestParameters,
+    ) -> ModelResponse:
+        self.captured_tools = model_request_parameters.builtin_tools
```

Comment:
Let's use `TestModel` instead, which already has `last_model_request_parameters` that we can read this info off of